File: client/src/types/case.ts
export interface JsonSchemaProperty {
  type: string;
  title: string;
  format?: string;
  enum?: string[];
}

export interface JsonSchema {
  type: string;
  properties: Record<string, JsonSchemaProperty>;
  required: string[];
}

export interface AuditType {
  id: string;
  name: string;
  slug: string;
  description: string | null;
  schema: JsonSchema;
  is_active: boolean;
  created_at: string;
}

export interface UserInfo {
  id: string;
  username: string;
  full_name: string;
  is_active: boolean;
}

export interface Case {
  id: string;
  case_number: number;
  title: string;
  description: string | null;
  audit_type_id: string;
  audit_type: AuditType | null;
  metadata: Record<string, unknown>;
  status: "open" | "active" | "closed";
  assigned_to_id: string | null;
  assigned_to: UserInfo | null;
  created_by_id: string;
  created_by: UserInfo | null;
  created_at: string;
  updated_at: string;
}

export interface CaseListResponse {
  items: Case[];
  total: number;
  offset: number;
  limit: number;
}

export interface CreateCaseRequest {
  title: string;
  audit_type_id: string;
  description?: string;
  metadata: Record<string, unknown>;
}

export interface UpdateCaseRequest {
  title?: string;
  description?: string;
  metadata?: Record<string, unknown>;
  status?: string;
  assigned_to_id?: string | null;
}

export interface CaseFilters {
  status?: string;
  audit_type_id?: string;
  assigned_to_id?: string;
  search?: string;
  offset?: number;
  limit?: number;
}

File: server/src/schemas/case.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from src.schemas.audit_type import AuditTypeRead
from src.schemas.user import UserRead


class CaseCreate(BaseModel):
    title: str = Field(max_length=200)
    description: str | None = None
    audit_type_id: uuid.UUID
    metadata: dict = Field(default_factory=dict)


class CaseUpdate(BaseModel):
    title: str | None = Field(default=None, max_length=200)
    description: str | None = None
    metadata: dict | None = None
    status: str | None = None
    assigned_to_id: uuid.UUID | None = None


class CaseRead(BaseModel):
    id: uuid.UUID
    case_number: int
    title: str
    description: str | None
    audit_type_id: uuid.UUID
    audit_type: AuditTypeRead | None = None
    metadata: dict = Field(validation_alias="metadata_")
    status: str
    assigned_to_id: uuid.UUID | None
    assigned_to: UserRead | None = None
    created_by_id: uuid.UUID
    created_by: UserRead | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


class CaseListResponse(BaseModel):
    items: list[CaseRead]
    total: int
    offset: int
    limit: int
